dhctl/pkg/kubernetes/actions/resources/watcher.go

```diff
@@ -0,0 +1,159 @@
+// Copyright 2024 Flant JSC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package resources
+
+import (
+	"context"
+	"fmt"
+	"io"
+
+	"github.com/deckhouse/deckhouse/dhctl/pkg/kubernetes/client"
+	"github.com/deckhouse/deckhouse/dhctl/pkg/log"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/client-go/dynamic/dynamicinformer"
+	"k8s.io/client-go/tools/cache"
+)
+
+type Watcher struct {
+	client  client.KubeClient
+	isReady func(object *unstructured.Unstructured) bool
+}
+
+func NewWatcher(client client.KubeClient) *Watcher {
+	return &Watcher{
+		client: client,
+	}
+}
+
+func (w *Watcher) WithIsReady(isReady func(object *unstructured.Unstructured) bool) *Watcher {
+	w.isReady = isReady
+	return w
+}
+
+func (w *Watcher) Watch(ctx context.Context, resourceID ResourceID, onReady func(*unstructured.Unstructured, bool)) (err error) {
+	if w.isReady == nil {
+		w.isReady = isReady
+	}
+
+	apiVersion, kind := resourceID.GroupVersionKind.ToAPIVersionAndKind()
+
+	groupVersionResource, err := w.client.GroupVersionResource(apiVersion, kind)
+	if err != nil {
+		return fmt.Errorf("error getting GroupVersionResource: %w", err)
+	}
+
+	informer := dynamicinformer.NewFilteredDynamicInformer(
+		w.client.Dynamic(),
+		groupVersionResource,
+		resourceID.Namespace,
+		0,
+		cache.Indexers{},
+		func(options *metav1.ListOptions) {
+			options.FieldSelector = fields.OneTermEqualSelector("metadata.name", resourceID.Name).String()
+		},
+	)
+
+	_, err = informer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(object interface{}) {
+				unstructuredObject := object.(*unstructured.Unstructured)
+
+				onReady(unstructuredObject, w.isReady(unstructuredObject))
+			},
+			UpdateFunc: func(_, object interface{}) {
+				unstructuredObject := object.(*unstructured.Unstructured)
+
+				onReady(unstructuredObject, w.isReady(unstructuredObject))
+			},
+			DeleteFunc: func(object interface{}) {
+				unstructuredObject := object.(*unstructured.Unstructured)
+
+				onReady(unstructuredObject, false)
```

Comment:
Why we need call `OnReady` in delete handler?

Reply:
Replaced with the `onDelete` handler